src/mscorlib/shared/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs

```diff
@@ -203,6 +221,92 @@ public override void Return(T[] array, bool clearArray = false)
             }
         }
 
+        public bool Trim()
+        {
+            int milliseconds = Environment.TickCount;
+            MemoryPressure pressure = GetMemoryPressure();
+
+            ArrayPoolEventSource log = ArrayPoolEventSource.Log;
+            if (log.IsEnabled())
+                log.BufferTrimPoll(milliseconds, (int)pressure);
+
+            foreach (PerCoreLockedStacks bucket in _buckets)
+            {
+                bucket?.Trim((uint)milliseconds, Id, pressure, _bucketArraySizes);
+            }
+
+            if (pressure == MemoryPressure.High)
+            {
+                // Under high pressure, release all thread locals
+                foreach (KeyValuePair<T[][], object> tlsBuckets in s_AllTlsBuckets)
+                {
+                    T[][] buckets = tlsBuckets.Key;
+                    for (int i = 0; i < NumBuckets; i++)
+                    {
+                        T[] buffer = buckets[i];
+                        buckets[i] = null;
+
+                        if (log.IsEnabled() && buffer != null)
+                        {
+                            log.BufferTrimmed(buffer.GetHashCode(), buffer.Length, Id);
```

Comment:
This might be wrong: it's possible a thread grabbed the buffer at the same time we did, in which case we'll report having trimmed a buffer even though we didn't.  I assume we're ok with that?

Reply:
Yes, unless you can think of a way to do this without incurring a hit. I'll clarify in comments.

Reply:
Actually, interlocked here isn't a bad idea since we're in cleanup.

Reply:
> Actually, interlocked here isn't a bad idea since we're in cleanup.

That might reduce the chances, but it won't remove them, unless you also used an interlocked on the rent path, which we don't want to do.

Reply:
What do you think? Worth the interlocked for a slimmer chance of false positive?

Reply:
Probably not, but if you wanted to, I'd suggest changing it to something like:
```C#
if (log.IsEnabled())
{
    T[] buffer = Interlocked.Exchange(buckets[i], null);
    if (buffer != null)
    {
        log.BufferTrimmed(buffer.GetHashCode(), buffer.Length, Id);
    }
}
else
{
    buckets[i] = null;
}
```